app/services/authorizer_cache.rb

```diff
@@ -1,14 +1,14 @@
 module AuthorizerCache
   def initialize_cache
-    @cache = HashWithIndifferentAccess.new do |h, k|
-      h[k] = HashWithIndifferentAccess.new
+    @cache = Hash.new do |h, k|
+      h[k] = {}
     end
   end
 
   def collection_cache_lookup(subject, permission)
-    collection = @cache[subject.class.to_s][permission] ||=
-      find_collection(subject.class, :permission => permission)
+    collection = @cache[subject.class.to_s][permission.to_s] ||=
+      find_collection(subject.class, :permission => permission).pluck(:id)
 
-    collection.include?(subject)
+    collection.include?(subject.id)
```

Comment:
Can you add a warning when too many items are loaded into memory? It can have a very meaningful explanation and big impact on how we are able to find these bottlenecks. Something like:

Authorizer cache loaded more than 1000 items, consider calling can? without cache to improve memory consumption.

I think this should be part of the patch which will backport, it will allow us to see more of these hotspots.

Reply:
good idea, i'll add a log but i think the threshold should be 100k - otherwise any user with many hosts will see this log all the time, and loading a 100k integer array is just 400k of memory used.

Reply:
Nice yeah 1000 was definitely too low, this looks good.

Reply:
Oh wait, this will issue INFO log on every check, I thought to have it in the block where `find_collection` is actually called to show it just once per subject check.

Reply:
We don't know the collection size inside `find_collection`, we only load the records after we get the collection.

Reply:
Sure but you are a programmer, you can solve this! :-)

Reply:
Does `.size` create a separate query that uses `COUNT()` or does it use the resulting set? Or does `.pluck()` return a simple array?

Reply:
Pluck return an array.